### PR TITLE
Mini-stats update with functionality and flexibility

### LINKED ISSUE #2416: Add draw call count to Mini Stats
One of the more common stat points for developers is draw calls. If we can add this to the mini stats, that would be great!

### PR DESCRIPTION
**Pull request by mvaligursky**
Mini-stats system has been updated to allow to render additional useful engine stats. It has also been made more configurable, and an engine example with example configuration has been added.

Fixes #2416

Screen Shot 2020-09-16 at 1 46 22 PM

engine example (when engine is release): https://playcanvas.github.io/#misc/mini-stats.html

Default configuration (to show what is configurable):

```
MiniStats.getDefaultOptions = function () {
    return {

        sizes: [
            { width: 100, height: 16, graphs: false },
            { width: 128, height: 32, graphs: true },
            { width: 256, height: 64, graphs: true }
        ],

        startSizeIndex: 0,

        colors: {
            graph0: new Color(0.7, 0.2, 0.2, 1),
            graph1: new Color(0.2, 0.7, 0.2, 1),
            graph2: new Color(0.2, 0.2, 0.7, 1),
            watermark: new Color(0.4, 0.4, 0.2, 1),
            background: new Color(0, 0, 0, 0.4)
        },

        cpu: {
            enabled: true,
            watermark: 33
        },

        gpu: {
            enabled: true,
            watermark: 33
        },

        stats: [
            {
                name: "Frame",
                stats: ["frame.ms"],
                decimalPlaces: 1,
                unitsName: "ms",
                watermark: 33
            }
        ]
    };
};

```

Example of customization:

```
        // set up options for mini-stats, start with the default options
        var options = pcx.MiniStats.getDefaultOptions();

        // configure sizes
        options.sizes = [
            { width: 100, height: 16, graphs: false },
            { width: 256, height: 32, graphs: true },
            { width: 500, height: 64, graphs: true }
        ];

        // when the application starts, use the largest size
        options.startSizeIndex = 2;

        // display additional counters
        // Note: for most of thees to report values, either debug or profiling engine build needs to be used.
        options.stats = [

            // frame update time in ms
            {
                name: "Update",
                stats: ["frame.updateTime"],
                decimalPlaces: 1,
                unitsName: "ms",
                watermark: 33
            },

            // total number of draw calls
            {
                name: "DrawCalls",
                stats: ["drawCalls.total"],
                watermark: 2000
            },

            // total number of triangles, in 1000s
            {
                name: "triCount",
                stats: ["frame.triangles"],
                decimalPlaces: 1,
                multiplier: 1 / 1000,
                unitsName: "k",
                watermark: 500
            },

            // number of materials used in a frame
            {
                name: "materials",
                stats: ["frame.materials"],
                watermark: 2000
            },

            // frame time it took to do frustum culling
            {
                name: "cull",
                stats: ["frame.cullTime"],
                decimalPlaces: 1,
                watermark: 1,
                unitsName: "ms",
            },

            // used VRAM, displayed using 2 colors - red for textures, green for geometry
            {
                name: "VRAM",
                stats: ["vram.tex", "vram.geom"],
                decimalPlaces: 1,
                multiplier: 1 / (1024 * 1024),
                unitsName: "MB",
                watermark: 100
            },

            // frames per second
            {
                name: "FPS",
                stats: ["frame.fps"],
                watermark: 60
            },

            // delta time
            {
                name: "dt",
                stats: ["frame.dt"],
                multiplier: 1000,
                decimalPlaces: 1,
                unitsName: "ms",
                watermark: 33
            }
        ];

        // create mini-stats system
        var miniStats = new pcx.MiniStats(app, options);

```

### CODE REVIEW

**Comment by slimbuck** [comment]
Running this locally I noticed the following:
- the background doesn't turn solid black with mouse over any more
- the watermark looks messy:
    - it has different heights in different graphs (2 vs 3 pixels)
    - it overwrites the graph and background (ideally it would blend with them)
- graph description and value text appears in different shades of grey (misaligned texels?)
- gaps between graphs should be 1 pixel (or configurable)

I am probably the worst offender when it comes to this, but shuffling the code around means it's virtually impossible to see what actual changes were made to this code. We should probably start submitting code-shuffle PRs separately from code changes.

**Comment by mvaligursky** [comment]
this is the commit where all changes were done after the file shuffling .. this makes it easy to see what has changed.
https://github.com/playcanvas/engine/pull/2428/commits/33a8f539ca895add5e9fca6fd2f04cff7fdc1706

**Comment by slimbuck** [comment]
Few more things:
- smallest (text-only) view is missing darkened background
- medium graph size should probably exclude watermark

File: extras/mini-stats/mini-stats.js [hunk 1]
```diff
@@ -442,22 +88,74 @@ function MiniStats(app) {
...
+
+        gpu: {
+            enabled: true,
+            watermark: 33
+        },
+
+        stats: [
+            {
+                name: "Frame",
+                stats: ["frame.ms"],
```
Comment (slimbuck) [inline extras/mini-stats/mini-stats.js:139]: It's unfortunate that this is now only available in profile and debug builds. Perhaps we can make an exception for his and #frame drawcalls and have them available in all builds? cc @willeastcott

**Comment by mvaligursky** [comment]
submitted improvements: 
- background when mouse is over is now darker (not fully dark as that hides rendering of the scene and does not make the graph more readable). It is configurable though.
- watermark is single pixel height
- space between graphs is smaller, and configurable
- default refresh changed to 500ms and made configurable (1s feels too slow)

other comments:
- now that the watermark is one pixel width .. it looks look good and is useful on smaller graph as well
- text has two brightness levels - numbers are brighter to stand out more even when mouse is not over the graph

**Comment by slimbuck** [comment]
THB I find the graphs are now less readable than before. When I'm not moused over them the text is not immediately readable (especially over a bright background) and when I do mouse over I find the background showing through to be distracting. I think these changes also take us further from the original design https://snapchat.quip.com/Ci9LAyRykPav.

**Comment by mvaligursky** [comment]
I made the changes as suggested. Visually I find it less pleasing, but I don't mind either way.

**Review by slimbuck — APPROVED** [review]

**Pull request merged**